spec/models/dialog_spec.rb

```diff
@@ -386,6 +386,38 @@
       end
     end
 
+    context "non-circular field associations" do
```

Comment:
With fewer database queries....
```ruby
context "non-circular field associations" do
  let(:dialog_tab) { FactoryGirl.create(:dialog_tab, :dialog => dialog) }
  let(:dialog_group) { FactoryGirl.create(:dialog_group, :dialog_tab => dialog_tab) }
  let(:dialog_field1) { FactoryGirl.create(:dialog_field, :dialog_group => dialog_group) }
  let(:dialog_field2) { FactoryGirl.create(:dialog_field, :dialog_group => dialog_group) }
  let(:dialog_field3) { FactoryGirl.create(:dialog_field, :dialog_group => dialog_group) }
  let(:dialog_field4) { FactoryGirl.create(:dialog_field, :dialog_group => dialog_group) }

  before { FactoryGirl.create(:dialog_field_association, :trigger => dialog_field1, :respond => dialog_field2 }

  it "adds error on simple circular reference" do
    FactoryGirl.create(:dialog_field_association, :trigger => dialog_field2, :respond_id => dialog_field1

    expect { dialog.save! }.to raise_error(ActiveRecord::RecordInvalid, /Dialog field associations cannot be circular/)
  end

  it "adds error on complex circular reference" do
    FactoryGirl.create(:dialog_field_association, :trigger => dialog_field2, :respond => dialog_field1)
    FactoryGirl.create(:dialog_field_association, :trigger => dialog_field1, :respond => dialog_field3)  # Is this one necessary?
    FactoryGirl.create(:dialog_field_association, :trigger => dialog_field1, :respond => dialog_field2)
    FactoryGirl.create(:dialog_field_association, :trigger => dialog_field2, :respond => dialog_field4)
    FactoryGirl.create(:dialog_field_association, :trigger => dialog_field4, :respond => dialog_field1)

    expect { dialog.save! }.to raise_error(ActiveRecord::RecordInvalid, /Dialog field associations cannot be circular/)
  end

  it "validates with non-circular reference" do
    expect { dialog.save! }.not_to raise_error
  end
end
```

Reply:
😭 this doesn't set the dialog tab to be on that dialog

Reply:
o_0 really?

Reply:
Oh, I see `FactoryGirl.build(:dialog)`

Reply:
Doesn't `build` just keep the instance in memory instead of persisting it? You still have to add the tabs.